src/Components/Web.JS/src/Boot.Server.Common.ts

```diff
@@ -83,42 +101,72 @@ export async function startServer(components: RootComponentManager<ServerCompone
   logger.log(LogLevel.Information, 'Blazor server-side application started.');
 
   jsInitializer.invokeAfterStartedCallbacks(Blazor);
+  resolve();
 }
 
-export function startCircuit(): Promise<boolean> {
-  if (!started) {
+export async function startCircuit(): Promise<boolean> {
+  if (!serverStartPromise) {
     throw new Error('Cannot start the circuit until Blazor Server has started.');
   }
 
-  if (circuit.didRenderingFail()) {
+  if (circuit && !circuit.isDisposedOrDisposing()) {
+    return true;
+  }
+
+  if (circuitStarting) {
+    return await circuitStarting;
+  }
+
+  await serverStartPromise;
+
+  if (circuit && circuit.didRenderingFail()) {
     // We can't start a new circuit after a rendering failure because the renderer
     // might be in an invalid state.
-    return Promise.resolve(false);
+    return false;
   }
 
-  if (circuit.isDisposedOrDisposing()) {
+  if (circuit && circuit.isDisposedOrDisposing()) {
     // If the current circuit is no longer available, create a new one.
     appState = discoverServerPersistedState(document) || '';
     circuit = new CircuitManager(circuit.getRootComponentManager(), appState, options, logger);
   }
 
   // Start the circuit. If the circuit has already started, this will return the existing
   // circuit start promise.
-  return circuit.start();
+  circuitStarting = circuit.start();
+  clearCircuitStarting(circuitStarting);
+  return circuitStarting;
+}
+
+async function clearCircuitStarting(circuitStartingPromise: Promise<boolean>) {
+  await circuitStartingPromise;
+  if (circuitStarting === circuitStartingPromise) {
+    circuitStarting = undefined;
+  }
```

Comment:
It took be a bit to understand but I get why this is necessary. Maybe a minor detail, but should the `circuitStarting` promise itself be the thing that sets `circuitStarting` to `undefined` before resolving? That way there's no possibility that someone awaiting `circuitStarting` will then call `startCircuit()` before this gets a chance to clear.

Reply:
If it's hard to understand there should be a comment here providing clarity

Reply:
`circuitStarting` is a module level variable, that gets assigned the result of `circuit.start()`, so I'd rather keep it as it is rather than have something in CircuitManager refer back to this file, or are you suggesting something else?

I've updated the code base with some comments for clarity.